File: backups/capacitor-sora-editor-1.3.13/android/sora-editor/editor/src/main/java/io/github/abc15018045126/sora/widget/snippet/variable/EditorBasedSnippetVariableResolver.java

package io.github.abc15018045126.sora.widget.snippet.variable;

import androidx.annotation.NonNull;

import io.github.abc15018045126.sora.text.ICUUtils;
import io.github.abc15018045126.sora.util.IntPair;
import io.github.abc15018045126.sora.widget.CodeEditor;

public class EditorBasedSnippetVariableResolver implements ISnippetVariableResolver {

    private final CodeEditor editor;

    public EditorBasedSnippetVariableResolver(@NonNull CodeEditor editor) {
        this.editor = editor;
    }

    @NonNull
    @Override
    public String[] getResolvableNames() {
        return new String[]{
                "TM_CURRENT_LINE", "TM_LINE_INDEX", "TM_LINE_NUMBER", "CURSOR_INDEX", "CURSOR_NUMBER",
                "TM_CURRENT_WORD", "SELECTION", "TM_SELECTED_TEXT"
        };
    }

    @NonNull
    @Override
    public String resolve(@NonNull String name) {
        switch (name) {
            case "TM_CURRENT_LINE":
            case "TM_LINE_NUMBER":
                return Integer.toString(editor.getCursor().getLeftLine() + 1);
            case "TM_LINE_INDEX":
                return Integer.toString(editor.getCursor().getLeftLine());
            case "CURSOR_INDEX":
                return Integer.toString(editor.getCursor().getLeft());
            case "CURSOR_NUMBER":
                return Integer.toString(editor.getCursor().getLeft() + 1);
            case "TM_CURRENT_WORD": {
                var text = editor.getText();
                var res = ICUUtils.getWordRange(text.getLine(text.getCursor().getLeftLine()), text.getCursor().getLeftColumn(), true);
                return text.getLine(text.getCursor().getLeftLine()).subSequence(IntPair.getFirst(res), IntPair.getSecond(res)).toString();
            }
            case "SELECTION":
            case "TM_SELECTED_TEXT": {
                var cursor = editor.getCursor();
                return editor.getText().subSequence(cursor.getLeft(), cursor.getRight()).toString();
            }
        }
        throw new IllegalArgumentException("Unsupported variable name:" + name);
    }
}

File: capacitor-sora-editor/android/sora-editor/editor/src/main/java/io/github/abc15018045126/sora/widget/snippet/variable/EditorBasedSnippetVariableResolver.kt
package io.github.abc15018045126.sora.widget.snippet.variable

import io.github.abc15018045126.sora.text.ICUUtils
import io.github.abc15018045126.sora.util.IntPair
import io.github.abc15018045126.sora.widget.CodeEditor

class EditorBasedSnippetVariableResolver(private val editor: CodeEditor) : ISnippetVariableResolver {

    override fun getResolvableNames(): Array<String> {
        return arrayOf(
            "TM_CURRENT_LINE", "TM_LINE_INDEX", "TM_LINE_NUMBER", "CURSOR_INDEX", "CURSOR_NUMBER",
            "TM_CURRENT_WORD", "SELECTION", "TM_SELECTED_TEXT"
        )
    }

    override fun resolve(name: String): String {
        return when (name) {
            "TM_CURRENT_LINE", "TM_LINE_NUMBER" -> (editor.cursor.leftLine + 1).toString()
            "TM_LINE_INDEX" -> editor.cursor.leftLine.toString()
            "CURSOR_INDEX" -> editor.cursor.left.toString()
            "CURSOR_NUMBER" -> (editor.cursor.left + 1).toString()
            "TM_CURRENT_WORD" -> {
                val text = editor.text
                val lineIndex = text.cursor.leftLine
                val line = text.getLine(lineIndex)
                val res = ICUUtils.getWordRange(line, text.cursor.leftColumn, true)
                line.subSequence(IntPair.getFirst(res), IntPair.getSecond(res)).toString()
            }
            "SELECTION", "TM_SELECTED_TEXT" -> {
                val cursor = editor.cursor
                editor.text.subSequence(cursor.left, cursor.right).toString()
            }
            else -> throw IllegalArgumentException("Unsupported variable name:$name")
        }
    }
}
